### ISSUE TITLE
[QA] SFT: Training loss fluctuate as steps grow, is it bad for downstream task?

### ISSUE THREAD

**Issue by weicheng59** [InternLM/InternLM#167]
### Describe the question.

I tried to SFT the Internlm-7b-chat using about 110k data of a specific task. 
Configs are default expect, zero1 7(using 7 gpu), micro_num 1, and micro_bsz 4. 
Using total steps of 4000 and 16000, I got following graphs showing training loss over time and learning rate overtime.
![7effee0d-910b-4486-b810-094653e7e740](https://github.com/InternLM/InternLM/assets/8345745/0fcff95f-9a1f-4a94-b4f0-da470b0ffacb)
![0bf8448b-b27a-43cd-85f3-0e53e4d025e8](https://github.com/InternLM/InternLM/assets/8345745/4e6df905-f8cb-4458-ab67-77972289db70)
As you may notice, the loss(in log scale) is fluctuating harder as steps grow. We wonder if this is normal considering loss is relative small 10e-3 to 10e-6.
Background information: We are trying to find the best time to stop training. It seemed like loss stop going lower around 13000 step but when we test the checkpoint from 2000 to 16000 using test cases. The best performer is actually at checkpoint 6000 in 16000 step run and checkpoint 4000 in 4000 step run. The best checkpoint and worst checkpoint only has a difference of 5% in terms of passing rate. 

Thanks for reading and great work on InternLM!

**Comment by ZwwWayne** [comment]
Hi @weicheng59 ,
From my perspective of view, the loss scale is already very small. So I guess it might be some overfitting issue. You can check the epoch number and whether there are some issues in your data.

**Comment by weicheng59** [comment]
Thanks for the reply. 
For epoch number, we choose 4000 step to get 1 epoch because we are using 7 card with micro_bsz 4 so each step will train 7x4 = 28 data and 4000*28 = 11.2k which roughly match the size of our dataset. However, as shown on the plot, log2's loss with deep green color at 4000 step show hard fluctuation.
Our guess for a relative small loss is there is high similarity in our training data. We are trying to finetune the model to do one specific task, converting natural language to xml format. So all our training data is like following example.
```
Instruction: 'Convert following text to myXml format',
Input: 'I love beach',
Output: '<Who>I</Who><Action>Love</Action><What>Beach</What>'
```
Our goal is not having model to answer question in a open manner so that may be that's the reason for small loss.

**Comment by ZwwWayne** [comment]
Understood. Then I think it might be fine. From my perspective view, the task is not so hard. So as long as the evaluation performance is satisfactory, it is fine.

**Comment by weicheng59** [comment]
Actually, we reached a bottle-neck of improving the pass rate to more than 52.5%. Right now we are taking two approaches:
1, improve training set
2, tuning the hyper-param
On the note of tuning the hyper-params, our experiment of 4 epoch vs 1 epoch show no improvement. I guess it's time to deep dive into training code and figure out a better loss calculation method for small loss value.

**Comment by ZwwWayne** [comment]
Yes you can try that.
